src/core/symbology-ng/qgsconcentriclegendsymbol.cpp

```diff
@@ -0,0 +1,132 @@
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsconcentriclegendsymbol.h"
+#include "qgssymbollayerutils.h"
+
+QgsConcentricLegendSymbol::QgsConcentricLegendSymbol( const QgsMarkerSymbol *symbol,
+    const QList< QPair< QString, double > > values,
+    const QgsUnitTypes::RenderUnit types,
+    const double maxSize,
+    const QgsDiagramSettings::SizeLegendType type )
+  : QgsMarkerSymbol( symbol->cloneLayers() )
+  , mValues( values )
+  , mMaxSize( maxSize )
+  , mType( type )
+{
+  for ( int i = mValues.size() - 1 ; i >= 0 ; i-- )
+  {
+    QPair< QString, double > value_size = mValues[i];
+    double size = value_size.second;
+
+    QgsMarkerSymbol *s = symbol->clone();
+    s->setSize( size );
+    s->setSizeUnit( types );
+    mSymbols << s;
+  }
+}
+
+QgsConcentricLegendSymbol::QgsConcentricLegendSymbol( const QgsConcentricLegendSymbol *const original )
+  : QgsMarkerSymbol( original->cloneLayers() )
+  , mSymbols( original->mSymbols )
+  , mValues( original->mValues )
+  , mMaxSize( original->mMaxSize )
+  , mType( original->mType )
+{
+}
+
+QgsConcentricLegendSymbol::~QgsConcentricLegendSymbol()
+{
+}
+
+
+QgsConcentricLegendSymbol *QgsConcentricLegendSymbol::clone() const
+{
+  return new QgsConcentricLegendSymbol( this );
+}
+
+void QgsConcentricLegendSymbol::drawPreviewIcon( QPainter *painter, QSize size, QgsRenderContext *customContext )
+{
+  QgsRenderContext context = customContext ? *customContext : QgsRenderContext::fromQPainter( painter );
+  context.setForceVectorOutput( true );
+  const double factor = context.scaleFactor();
+  QPainter *const p = context.painter();
+  p->save();
+
+  int max_label_size = 0;
+  Q_FOREACH ( LabelValue lv, mValues )
+  {
+    max_label_size = std::max( max_label_size, p->fontMetrics().width( lv.first ) );
+  }
+  p->translate( -max_label_size - 3 * factor, 0 );
+
+
+  Q_FOREACH ( QgsMarkerSymbol *s, mSymbols )
+  {
+    p->save();
+    switch ( mType )
+    {
+      case QgsDiagramSettings::ConcentricBottom:
+        p->translate( 0, ( mMaxSize - s->size() ) * factor / 2 );
+        break;
+      case QgsDiagramSettings::ConcentricCenter:
+        break;
+      case QgsDiagramSettings::ConcentricTop:
+        p->translate( 0, ( s->size() - mMaxSize ) * factor / 2 );
+        break;
+      default:
+        throw QgsException( QStringLiteral( "Wrong type value" ) );
```

Comment:
I don't know, but it should never happens.
Is it better to:
- use an assert `assert(false)`
- consider it as `ConcentricTop`
?